File: C:C++/Sort/C/Bubble_sort.c
#include <stdio.h>

// Function to swap two integers
void swap(int *xp, int *yp) {
    int temp = *xp;
    *xp = *yp;
    *yp = temp;
}

// Function to perform bubble sort
void bubbleSort(int arr[], int n) {
    // Iterate over the array from the end to the beginning
    for (int i = n - 1; i > 0; i--) {
        int exchange = 0; // Flag to check if any swaps are made
        // Iterate over the unsorted portion of the array
        for (int j = 0; j < i; j++) {
            // Swap adjacent elements if they are in the wrong order
            if (arr[j] > arr[j + 1]) {
                swap(&arr[j], &arr[j + 1]); // Call swap function
                exchange = 1; // Set the exchange flag to true
            }
        }
        // If no swaps are made in a pass, the array is already sorted
        if (!exchange) {
            break;
        }
    }
}

int main() {
    int arr[] = {5, 4, 3, 2, 1}; // Input array
    int n = sizeof(arr) / sizeof(arr[0]); // Calculate the size of the array
    bubbleSort(arr, n); // Call bubble sort function
    // Print the sorted array
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
    return 0;
}

File: Python/Data_Structure_and_Algorithm/Bubble_sort.py
# Bubble_sort_v1
def bubble_sort(a_list):
    # Iterate over the list in reverse order
    # Starting from len(a_list) - 1 down to 1 (inclusive) because the last element is already in its final position
    for i in range(len(a_list) - 1, 0, -1):
        # Iterate over the list from index 0 to i - 1
        # Compare each adjacent pair of elements and swap if necessary
        for j in range(i):
            # If the current element is greater than the next element, swap them
            if a_list[j] > a_list[j + 1]:
                # Swap elements
                temp = a_list[j]
                a_list[j] = a_list[j + 1]
                a_list[j + 1] = temp


# Bubble_sort_v2
def bubble_sort_short(a_list):
    # Iterate over the list in reverse order
    # Starting from len(a_list) - 1 down to 1 because the last element is already in its final position
    for i in range(len(a_list) - 1, 0, -1):
        exchanges = False  # Flag to track if any exchanges were made in this iteration
        # Iterate over the list from index 0 to i - 1
        # Compare each adjacent pair of elements and swap if necessary
        for j in range(i):
            # If the current element is greater than the next element, swap them
            if a_list[j] > a_list[j + 1]:
                # Swap elements
                a_list[j], a_list[j + 1] = a_list[j + 1], a_list[j]
                exchanges = True  # Set the flag to True indicating an exchange was made
        # If no exchanges were made in this iteration, the list is already sorted, so break out of the loop
        if not exchanges:
            break
